### PR TITLE
[7.x] Implement anonymous components

### PR DESCRIPTION
**Pull request by driesvints**
This PR introduces class-less components à la Blade X. What it does under the hood is make use of a `ClassLessComponent` class (wait whut?) to pipe through all the attributes as data to the component. This way you can do this:

**welcome.blade.php**

```
<x-avatar size="100" />
```

**components/avatar.blade.php**

```
@php($user = $user ?? Auth::user())
@php($size = $size ?? 50)

<img
    class="inline-block rounded-full"
    src="{{ $user->gravatarUrl($size) }}"
    width="{{ $size }}"
    height="{{ $size }}"
/>
```

All without a component class. This will be super useful for very simple small components and composition where you don't always need a class for.

### Update

Decided to rename `ClassLessCompnent` to `AnonymousComponent`.

### CODE REVIEW

File: src/Illuminate/View/ClassLessComponent.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
+<?php
+
+namespace Illuminate\View;
+
+class ClassLessComponent extends Component
```
Comment (driesvints) [inline src/Illuminate/View/ClassLessComponent.php]: I'm not particularly happy with this class name so any suggestion for a better one is welcome.

File: src/Illuminate/View/ClassLessComponent.php [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+     */
+    protected $data = [];
+
+    /**
+     * Create a new class-less component instance.
+     *
+     * @param  array  $data
+     * @return void
+     */
+    public function __construct($data)
```
Comment (driesvints) [inline src/Illuminate/View/ClassLessComponent.php]: I wanted to pass the view as a separate parameter but couldn't immediately figure out how. Will need to be cracked in the `componentString` method of the `ComponentTagCompiler` class.

File: src/Illuminate/View/ClassLessComponent.php (hunk 1 above)
Comment (GrahamCampbell) [reply src/Illuminate/View/ClassLessComponent.php]: `AnonymousComponent`?

File: src/Illuminate/View/Compilers/ComponentTagCompiler.php [hunk 3]
```diff
@@ -153,7 +153,19 @@ protected function componentString(string $component, array $attributes)
...
+
+            $class = ClassLessComponent::class;
+
+            $parameters = "'data' => [".$this->attributesToString($data->all()).']';
+        } else {
+            $parameters = $this->attributesToString($data->all());
+        }
+
+        return " @component('{$class}', [".$parameters.'])
 <?php $component->withAttributes(['.$this->attributesToString($attributes->all()).']); ?>';
```
Comment (driesvints) [inline src/Illuminate/View/Compilers/ComponentTagCompiler.php]: I wanted to use `var_export` for this instead but unfortunally the blade compiler doesn't plays well with the old array syntax.

File: src/Illuminate/View/Compilers/ComponentTagCompiler.php [hunk 4]
```diff
@@ -169,11 +181,11 @@ protected function componentClass(string $component)
...
         }
 
-        if (! class_exists($class = $this->guessClassName($component))) {
-            throw new InvalidArgumentException("Unable to locate class for component [{$component}].");
+        if (class_exists($class = $this->guessClassName($component))) {
+            return $class;
         }
 
-        return $class;
+        return $component;
```
Comment (driesvints) [inline src/Illuminate/View/Compilers/ComponentTagCompiler.php]: We could also do a view exists check here possible.

File: src/Illuminate/View/ClassLessComponent.php (hunk 1 above)
Comment (driesvints) [reply src/Illuminate/View/ClassLessComponent.php]: Yeah that's better. Thanks. I'm gonna wait a bit more to see if there are any more suggestions.

File: src/Illuminate/View/ClassLessComponent.php (hunk 2 above)
Comment (driesvints) [reply src/Illuminate/View/ClassLessComponent.php]: Solved this 👍

File: src/Illuminate/View/ClassLessComponent.php (hunk 1 above)
Comment (ryangjchandler) [reply src/Illuminate/View/ClassLessComponent.php]: +1 for AnonymousComponent, seems for language-fitting.

File: src/Illuminate/View/Compilers/ComponentTagCompiler.php (hunk 4 above)
Comment (driesvints) [reply src/Illuminate/View/Compilers/ComponentTagCompiler.php]: Fixed.

File: src/Illuminate/View/ClassLessComponent.php (hunk 1 above)
Comment (driesvints) [reply src/Illuminate/View/ClassLessComponent.php]: Renamed to `AnonymousComponent`. Thanks!

File: src/Illuminate/View/Compilers/ComponentTagCompiler.php (hunk 3 above)
Comment (GrahamCampbell) [reply src/Illuminate/View/Compilers/ComponentTagCompiler.php]: You can get short array syntax by pretty printing using php-parser, I think?

Comment (driesvints) [reply src/Illuminate/View/Compilers/ComponentTagCompiler.php]: Maybe overkill. Feel free to pr afterwards if you want.

Comment (GrahamCampbell) [reply src/Illuminate/View/Compilers/ComponentTagCompiler.php]: Yeh, definitely overkill. We can mark this as resolved.

**Review by GrahamCampbell — APPROVED** [review]

File: src/Illuminate/View/Compilers/ComponentTagCompiler.php [hunk 5]
```diff
@@ -153,7 +155,17 @@ protected function componentString(string $component, array $attributes)
 
         [$data, $attributes] = $this->partitionDataAndAttributes($class, $attributes);
 
-        return " @component('{$class}', [".$this->attributesToString($data->all()).'])
+        // If the component doesn't exists as a class we'll assume it's a class-less component
+        // and pass the component as a view parameter to the data collection bag.
+        if (! class_exists($class)) {
```
Comment (GrahamCampbell) [inline src/Illuminate/View/Compilers/ComponentTagCompiler.php]: The only thing that worries me is how to distinguish between a user making a typo in a class name, and actually wanting to not have a class.

Comment (driesvints) [reply src/Illuminate/View/Compilers/ComponentTagCompiler.php]: That's a really big edge case imo. Plus I guess you'll immediately see it?

**Comment by taylorotwell** [comment]
With class based components you can use colons to indicate directories like:

`<x-inputs:text />`

Will resolve to `App\View\Components\Inputs\Text.php` by default. Do these anonymous components allow you to do that so `<x-inputs:text />` as an anonymous component would resolve to `views/components/inputs/text.blade.php`?

**Comment by ryangjchandler** [comment]
> With class based components you can use colons to indicate directories like:
> 
> `<x-inputs:text />`
> 
> Will resolve to `App\View\Components\Inputs\Text.php` by default. Do these anonymous components allow you to do that so `<x-inputs:text />` as an anonymous component would resolve to `views/components/inputs/text.blade.php`?

At a quick glance, I'm not sure it will since the `AnonymousComponent` class is just receiving the FQN of what would be the defined component's class. The `view()` method on the `AnonymousComponent` class would need to handle the magic, transform the class name to a valid Blade path, something like below.

```php
function view()
{
    return collect(explode('\\', Str::after($this->view, '\\View\\Components\\')))->map(function ($segment) {
        return Str::kebab($segment);
    })->implode('.');
}
``` 

If `$this->view` has been set to `App\\View\\Components\\Test\\Component', the output would be `test.component`.

**Comment by garygreen** [comment]
Shouldn't this really live in userland / as a package? Personally I'm not much of a fan of Blade X design paradigm - it's very opinionated and Vue-like.

It's useful to take a step back and remember that PHP itself is a templating language and Blade is a relatively simple abstraction on top of it. Having a new paradigm for creating components in Laravel just further complicates and pushes the hurdles forwards, particularly for Laravel newcomers for relatively little gain imo.

**Comment by ryangjchandler** [comment]
@garygreen, I think it's perfectly okay to have this feature in core. Using the BladeX style / Vue-like tags is not required, as these tags get compiled down to regular Blade directives anyway, you could achieve the same outcome using the Blade directives. 

The custom element style tags and class based components are just a nicety that a lot of people, especially those currently using the BladeX package or laravel-view-components packages will like.

**Comment by garygreen** [comment]
This could be useful for some developers but I'm ultimately questioning whether it **needs** to be in core - why not just `composer require` the package you need? Although you may find it useful I think it's worth considering the Laravel ecosystem as a whole, particulary for newcomers - with additional ways of doing the same sort of thing (rendering components) with a slightly different syntax just further pushes the learning curve for relatively little gain.

**Comment by ryangjchandler** [comment]
Im not saying it's needed in the core, but it's definitely a nice touch. The same logic, that it's confusing could be applied to other areas of Laravel. If you're using Notification to just send an email, why not use a Mailable? They would both ultimately have the same outcome, both can be queued. Why use middleware to prevent users without particular permissions accessing routes, you could use the authorize() method of a FormRequest class. 

The reason I believe in offering multiple approaches is for developers, newcomers or not, to find the way that suits them best, without them having to install third-party packages or resort to custom solutions. If functionality can be provided in core, there's very little time spent getting up and running. You simply put a piece of code in. No need to require a package, publish vendor related resources, modify config files. 

Not disagreeing with you, but I'm also not agreeing with you. Somewhere in the middle, if that's allowed 🤦‍♂️

**Review by vkarampinis — APPROVED** [review]

**Comment by taylorotwell** [comment]
I'm also curious how something like <x-anon-component wire:model="foo"> would work? Since wire:model is not a valid PHP variable name? It seems like you would actually have to put everything in an attribute bag instead of making them variables?

**Comment by taylorotwell** [comment]
@calebporzio suggested perhaps a @props directive at the top of anon components that specifies which variables should be extracted from the attribute bag and made real variables...

image

**Comment by garygreen** [comment]
> I'm also curious how something like <x-anon-component wire:model="foo"> would work? Since wire:model is not a valid PHP variable name

This is why this PR and Blade X being appealing is so strange to me. That is completely valid in html/js and you could access that attribute via `el.getAttribute('wire:model')` in PHP it doesn't make sense and your wrapping your component logic into a HTML tag and then destructing it back into PHP - why?

In Vue this is the main way of including components and exposing props - it solves a complicated problem on the frontend, but in PHP that simply isn't a problem and doesn't solve anything. It just creates an unnescary intermediary.

Anyway, that's just my take on it. I'm sure this comment will get deleted. 😋

**Comment by taylorotwell** [comment]
Actually it is solving a lot of problems - just because you don't see them doesn't mean they don't exist. Comments get deleted if they are off topic.

File: src/Illuminate/View/Compilers/ComponentTagCompiler.php [hunk 6]
```diff
@@ -320,7 +344,11 @@ protected function parseBindAttributes(string $attributeString)
     protected function attributesToString(array $attributes)
     {
         return collect($attributes)
-                ->map(function (string $value, string $attribute) {
+                ->map(function ($value, string $attribute) {
+                    if (is_array($value)) {
+                        $value = '['.$this->attributesToString($value).']';
```
Comment (taylorotwell) [inline src/Illuminate/View/Compilers/ComponentTagCompiler.php]: What is this needed for?

**Comment by garygreen** [comment]
The alternative way of writing that would be:

```php
@include('map', ['target' => 'map', 'class' => 'mb-4 border-2'])
```

Within the map component:

```
<div class="{{ $class }} flex items-center">
  Do something with {{ $target }}
</div>
```

My point is that with Blade X your using HTML as a "intermediary" to pass information onto a component - this isn't strictly "solving" anything. If anything it introduces complexity into how those components are parsed, resolved, variable naming (as you discovered) etc.

I'm genuinely finding it difficult to see the benefit's of Blade X / Anonymous Components beyond just being opinionated syntax sugar.

**Comment by taylorotwell** [comment]
@garygreen OK, now take your example and support ANY arbitrary HTML element. How would you go about it?

```php
<x-input name="name" class="mb-4" wire:model.debounce.1000ms="name" required autofocus />
```

**Comment by garygreen** [comment]
What would that actually render to?

**Comment by taylorotwell** [comment]
I posted what it would render to?

**Comment by taylorotwell** [comment]
image

image

image

File: src/Illuminate/View/Compilers/ComponentTagCompiler.php [hunk 7]
```diff
@@ -350,7 +350,7 @@ protected function parseBindAttributes(string $attributeString)
     protected function attributesToString(array $attributes)
     {
         return collect($attributes)
-                ->map(function ($value, string $attribute) {
+                ->map(function (string $value, string $attribute) {
```
Comment (GrahamCampbell) [inline src/Illuminate/View/Compilers/ComponentTagCompiler.php]: Why do we have scalar types here?

Comment (taylorotwell) [reply src/Illuminate/View/Compilers/ComponentTagCompiler.php]: I'm not really opposed to using them. I'm just not normally in the habit from so many years of them not being available.

Comment (GrahamCampbell) [reply src/Illuminate/View/Compilers/ComponentTagCompiler.php]: We could delete them here. The code will throw an exception if they are not castable to a string on the next line anyway.

Comment (driesvints) [reply src/Illuminate/View/Compilers/ComponentTagCompiler.php]: Let's do that in a separate pr :)

**Comment by garygreen** [comment]
Ah gotcha. I can see how that is useful, does remind me of `Form::input()` or some kind of `{{ $input->text }}` type builder although one benefit of it being it keeps everything in HTML whilst still allowing you to hook into it and alter how it's finally rendered.

Is this capable of handling a body of text?

```php
<x-textarea>blah</x-textarea>
```

How do you supply dynamic options to it? Looking at the docs of BladeX it looks like it would be:

```html
<x-option :checked="$booking->involvesUser($user)" :text="$user->name">
```

Also does the blade compiler put the "compiled" version of the code in the view or does it have to create the anonymous class object every time for every component included on the page? If you had thousands of inputs it may cause slower rendering compared to a custom optimized blade directive:

```php
/* May not look as pretty but one benefit of this
/* approach is it keeps in the realm of PHP so
/* gives you a lot of dynamic control */
@x('option', ['checked' => $booking->involvesUser($user), 'text' => $user->name])
```

Anyway, I'm warming to it. 👍 Especially if it's fast in rendering many micro components.

**Comment by driesvints** [comment]
> Is this capable of handling a body of text?

Yeah slots just work as normal.

> How do you supply dynamic options to it? Looking at the docs of BladeX it looks like it would be:

You can do that here as well.

> Also does the blade compiler put the "compiled" version of the code in the view or does it have to create the anonymous class object every time for every component included on the page? If you had thousands of inputs it may cause slower rendering compared to a custom optimized blade directive:

This isn't a problem because views are compiled and cached.

**Comment by taylorotwell** [comment]
Yes, it handles slots just like normal components.

Yes, it handles dynamic options exactly like Blade X (same syntax there).

Yes, it creates a class to use a component.

**Comment by taylorotwell** [comment]
For reference... rendering 1000 components on a screen takes about 50ms on my machine. So, yes it can be "slow". Rendering 1000 `@include` statements takes about 30ms.

A more "typical" page rendering 100 components takes about 5-6ms to render (vs about 4ms using `@include`).

**Comment by mojtabaahn** [comment]
$attributes return nothing currently when using anonymous components..

**Comment by mojtabaahn** [comment]
Aright nvm, It used to not working on 7.x-dev but in 7.0.2 is working fine!

**Comment by devfaysal** [comment]
If I have a package that has some anonymous components the only way I can make them available is to create a class for them and register them individually. Is there some (undocumented) way of loading from other locations?
Any idea @driesvints @taylorotwell

**Comment by driesvints** [comment]
@devfaysal no: https://laravel.com/docs/7.x/blade#components

**Comment by devfaysal** [comment]
> @devfaysal no: https://laravel.com/docs/7.x/blade#components

Thanks for quick reply. I created an issue #32154  about it.

**Pull request merged**